### PR TITLE
refactor: use batch confirmation that was upstreamed to eigenda client

### PR DESCRIPTION
**Pull request by samlaf**
This PR depends on https://github.com/Layr-Labs/eigenda/pull/821 being merged first.

Main simplification that this allows is getting rid of the very complex retry logic that we had where an inner functionw as returning an error that was being caught by a very outward for loop, which was hard to understand. This PR makes that logic more local, but I opted to still keep the check in eigenda-proxy as well as an assert/safety guarantee. Think we also need it in the Get route... but not super familiar with that code path cc @epociask to make sure logic in this PR still works for that.

## Fixes Issue

Fixes #

### CODE REVIEW

File: go.mod
Comment (samlaf) [inline go.mod:299]: TODO

File: go.mod
Comment (samlaf) [inline go.mod:11]: TODO: merge https://github.com/Layr-Labs/eigenda/pull/821 before this PR

Comment (samlaf) [reply go.mod:11]: Also notttt....... sure why the ci workflows are all broken. Can't find batchHeader... at lines that don't even exist. lint test and build work fine locally for me so I'm confused. Wondering if it doesn't have to do with this reference somehow..

File: verify/cert.go
Comment (samlaf) [inline verify/cert.go:67]: Is this true? If so I think we should get rid of this check but document exactly why we are getting rid of it. Would have saved me a lot of time if this had been documented in the initial code.

**Review by bxue-l2 — DISMISSED** [review]

File: verify/cert.go
Comment (epociask) [inline verify/cert.go:56]: Knit
```suggestion
func (cv *CertVerifier) verifyBatchConfirmedOnChain(
```

File: verify/cert.go
Comment (epociask) [inline verify/cert.go:72]: why would we ever use two nodes? or is the idea to have some resiliency to transient failures where the provider could be unreliable due to poor configuration?

File: verify/cert.go
Comment (epociask) [inline verify/cert.go:98]: knit - there is an equality func in utils: https://github.com/Layr-Labs/eigenda-proxy/blob/main/utils/utils.go#L31-L43

File: verify/cert.go
Comment (epociask) [inline verify/cert.go:153]: knit
```suggestion
		return [32]byte{}, fmt.Errorf("error calling EigenDAServiceManager.BatchIdToBatchMetadataHash: %w", err)
```

File: verify/cert.go
Comment (epociask) [inline verify/cert.go:132]: assuming this is encoded to the raw params like [`finalized, `false`] where the boolean indicates that we aren't collecting txs for the block? We may wanna provide more explicit detailing of this value for easier readability long-term

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go:67]: Interesting. But I'm not even sure, because the case I'm describing is this:
0. (assumption: we wait for 5 block confirmation)
1. batcher confirms batch onchain at block 1000
2. eigenda-client wait for 5 blocks and returns a BlobStatus
3. proxy receives this one block later at block 1006, and the batch has actually been reorged and included in block 1001 instead.
4. In this case the check highlighted here would fail, even though the block is actually confirmed with depth 5 (the next assert below around line 70 would catch if that was not the case)

For this reason I think this check is actually wrong here, and I still think we should remove this check. Does this make sense?

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go]: [b1550e3](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/b1550e38518627449974c2cd38ed6c389389bc33)

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go:72]: Both eigenda-client and proxy dial the endpoint individually, establishing independent connections. If the nodes are behind a load-balancer, then both connections could actually be connected to different eth-nodes, its very frequent that rpc provider nodes are not in sync (bit me very hard once) for some dumb reason (web3 standards...).

Even if we injected the dialed connection from the proxy into the eigenda-client, the underlying tcp connection could drop in between the eigenda-client returning and proxy making its call, which would reconnect to another node, with some effect.

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go:98]: Didn't know! But checked and that implementation is literally the same as [slices.Equal](https://pkg.go.dev/golang.org/x/exp/slices#Equal) though. I'd prefer sticking with stdlib (even if in exp package) stuff when we can.

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go:153]: Disagree here. Read https://github.com/uber-go/guide/blob/master/style.md#error-wrapping last week which makes a lot of sense, been trying to use this convention as much as possible.

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go:134]: [7069e13](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/7069e13e252623e6e9b163e7e8b628e7c662e2bb)

File: flags/eigendaflags/cli.go
Comment (samlaf) [reply flags/eigendaflags/cli.go:24]: Thought about this but not sure I like it. Perhaps we could do something like what the json-rpc does (where you can either pass a block number, or a string like "finalized" or "latest")?
```
cli.StringFlag{
    Name:  "confirmation-mode",
    Usage: "Confirmation mode: <number of blocks> or 'finalized'",
    Value: "finalized",
}
```

Comment (samlaf) [reply flags/eigendaflags/cli.go]: Made the change: [fbcf650](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/fbcf650275252985cc1bd4c403cea59fbc9cf897)

**Comment by samlaf** [comment]
@epociask I'll let you review my answers before rebasing and fixing conflicts, so that the commit links still work.

File: flags/eigendaflags/cli.go
Comment (epociask) [inline flags/eigendaflags/cli.go:163]: should probably rm this line

File: flags/eigendaflags/cli.go
Comment (epociask) [inline flags/eigendaflags/cli.go:191]: Is there any potential value using geth's json rpc unmarshaling directly to get cross compatibility with other commitment levels? (e.g, `latest`, `safe`)  without having to redefine types?

https://github.com/ethereum/go-ethereum/blob/80bdab757dfb0f6d73fb869d834979536fe474e5/rpc/types.go#L73-L112

File: verify/cert.go
Comment (epociask) [reply verify/cert.go:72]: makes sense and was thinking that too - got burnt in the past using a node cluster with an ALB ingress where subsequent RPC queries could fail due to nodes being out of sync. Ik OP clusters typically use [proxyd](https://github.com/ethereum-optimism/infra/tree/main/proxyd) to mitigate this from happening. from what I remember the connection would be held with the ALB directly with different nodes being queried per each new request on the client --> ALB connection. Feel free to resolve this comment!

File: verify/cert.go
Comment (epociask) [reply verify/cert.go:98]: understood - lets cut an issue to remove that utils function entirely then

File: verify/cert.go
Comment (epociask) [reply verify/cert.go:72]: Actually on second thought - isn't this a bit naive when we have access to the `L1ReferenceBlock` number at which the batch should've confirmed? Also we're adding a 12 second cool-off which is aligned with the block production interval on Ethereum but not syncing

Comment (samlaf) [reply verify/cert.go:72]: proxyd doesn't seem to solve this problem, they say: "we consider healthy is... latest block lag ≤ configurable threshold". Way to solve this is to use sticky sessions.

Not sure I understand your last comment. Is https://github.com/Layr-Labs/eigenda-proxy/pull/192#discussion_r1818085482 related to what you are talking about?

File: flags/eigendaflags/cli.go
Comment (samlaf) [reply flags/eigendaflags/cli.go]: yep thanks, removed

File: flags/eigendaflags/cli.go
Comment (samlaf) [reply flags/eigendaflags/cli.go:190]: Maybe, but they take a byte as input, and accept many more strings. For now let's keep it like this.

File: verify/cert.go
Comment (bxue-l2) [reply verify/cert.go:67]: EigenDA batcher detects and updates confirmation number, if it detects reorg, https://github.com/Layr-Labs/eigenda/blob/master/disperser/batcher/finalizer.go#L198

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go:72]: Talked offline. Let's update to retrying every 2 seconds for the same amount of time (60 seconds).

File: go.mod
Comment (samlaf) [reply go.mod]: merged + updated: [6a25446](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/6a25446508b5e9e4a89f2fabd0027d1b7f739e4b)

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go]: [55edfae](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/55edfaefb4297be0dbd31e6251db691b951560a4)

File: verify/cert.go
Comment (samlaf) [reply verify/cert.go]: [0c54503](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/0c545036bcf8f27cb9dac9882aa8649fd508a659)

**Review by epociask — APPROVED** [review]
LGTM

**Pull request merged**
